test/extended/router/grpc-interop.go

```diff
@@ -3,177 +3,206 @@ package router
+			image, err := getCanaryImage(oc)
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("Creating grpc-interop test service")
+			err = oc.Run("new-app").Args("-f", grpcServiceConfigPath, "-p", "IMAGE="+image).Execute()
 			o.Expect(err).NotTo(o.HaveOccurred())
-			e2e.ExpectNoError(e2epod.WaitForPodRunningInNamespaceSlow(oc.KubeClient(), "grpc-interop", oc.KubeFramework().Namespace.Name))
 
-			g.By("Discovering the set of supported test cases")
-			ns := oc.KubeFramework().Namespace.Name
-			output, err := grpcInteropExecClientShellCmd(ns, gRPCInteropTestTimeout, "/workdir/grpc-client -list-tests")
+			g.By("Waiting for grpc-interop pod to be running")
+			e2e.ExpectNoError(e2epod.WaitForPodRunningInNamespaceSlow(oc.KubeClient(), "grpc-interop", oc.Namespace()), "grpc-interop backend server pod not running")
+
+			// certificate start and end time are very
+			// lenient to avoid any clock drift between
+			// between the test machine and the cluster
+			// under test.
+			notBefore := time.Now().Add(-24 * time.Hour)
+			notAfter := time.Now().Add(24 * time.Hour)
+
+			// Generate crt/key for routes that need them.
+			_, tlsCrtData, tlsPrivateKey, err := certgen.GenerateKeyPair(notBefore, notAfter)
 			o.Expect(err).NotTo(o.HaveOccurred())
-			testCases := strings.Split(strings.TrimSpace(output), "\n")
-			o.Expect(testCases).ShouldNot(o.BeEmpty())
-			sort.Strings(testCases)
 
-			grpcInteropExecInClusterRouteTests(oc, testCases, gRPCInteropTestCaseIterations)
-			grpcInteropExecInClusterServiceTests(oc, testCases, gRPCInteropTestCaseIterations)
-			grpcInteropExecOutOfClusterRouteTests(oc, testCases, gRPCInteropTestCaseIterations)
+			derKey, err := certgen.MarshalPrivateKeyToDERFormat(tlsPrivateKey)
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			pemCrt, err := certgen.MarshalCertToPEMString(tlsCrtData)
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			shardFQDN := oc.Namespace() + "." + defaultDomain
+
+			g.By("Creating routes to test for gRPC interoperability")
+			err = oc.Run("new-app").Args("-f", grpcRoutesConfigPath,
+				"-p", "DOMAIN="+shardFQDN,
+				"-p", "TLS_CRT="+pemCrt,
+				"-p", "TLS_KEY="+derKey,
+				"-p", "TYPE="+oc.Namespace()).Execute()
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("Creating a test-specific router shard")
+			shardConfigPath, err = shard.DeployNewRouterShard(oc, 10*time.Minute, shard.Config{
+				FixturePath: grpcRouterShardConfigPath,
+				Domain:      shardFQDN,
+				Type:        oc.Namespace(),
+			})
+			o.Expect(err).NotTo(o.HaveOccurred(), "new router shard did not rollout")
+
+			// Shard is using a namespace selector so
+			// label the test namespace to match.
+			g.By("By labelling the namespace")
+			err = oc.AsAdmin().Run("label").Args("namespace", oc.Namespace(), "type="+oc.Namespace()).Execute()
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("Getting LB service")
+			shardService, err := getRouterService(oc, 10*time.Minute, "router-"+oc.Namespace())
```

Comment:
10 minutes to get the ingress controller's service seems high.

Reply:
Dropped to 5 mins in https://github.com/openshift/origin/pull/25897/commits/1f2980383228d53218a0f8ba266ed89d7b976734.

@sgreene570 if you have no other requests I propose to squash and we can move this forward.

Reply:
Ack. Squash away!

Reply:
Done.